### PR TITLE
net: context: Set target network interface in send if needed

### PR DESCRIPTION
**Pull request by jukkar**
If we are sending a network packet and if the remote address
is not set in the context (which means that bind() has not
been called), then we must set the target network interface
to a proper value. This is required if we have multiple network
interfaces in which case the packet would be always assigned to
first network interface if remote address is unspecified.

Noticed this while debugging #32750. This issue has been undetected earlier but after commit 157c11d6e7072b5cf1ae8816a5673147c0e587f5 we started to return `-ENETDOWN` if network interface is down. In my tests, I had several network interfaces, but sometimes when using the `net udp` command, the sending failed because the first interface was down. In that case, the packet was meant to be sent to the second interface so it was quite mysterious why that happened.
The fact that the network packet was sent to wrong interface might explain some weird routing issues with VLANs.

### CODE REVIEW

**Review by rlubos — APPROVED** [review]

**Review by hubertmis — COMMENTED** [review]
The change looks correct to me. However, I don't know Zephyr network stack well enough to assess if proposed interface handling is correct.

I think by checking if remote address in given context is unspecified, you intended to check if given context is not connected (remote address is set in `net_context_connect()`) and in such circumstances you override interface that potentially have been set during binding process. If I see correctly, interface is used for net contexts only in send path. In such circumstances, overriding seems correct.

I'm wondering why interface is set for context during binding phase, that is part of receive path. But it seems out of scope of this patch.

File: subsys/net/ip/net_context.c [hunk 1]
```diff
@@ -1462,6 +1462,14 @@ static int context_sendto(struct net_context *context,
 		if (net_ipv6_is_addr_unspecified(&addr6->sin6_addr)) {
 			return -EDESTADDRREQ;
 		}
+
+		if (addr6 && memcmp(&net_sin6(&context->remote)->sin6_addr,
```
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/ip/net_context.c:1466]: Good point, I will change this.

File: subsys/net/ip/net_context.c [hunk 2]
```diff
@@ -1462,6 +1462,14 @@ static int context_sendto(struct net_context *context,
...
+				    net_ipv6_unspecified_address(),
+				    sizeof(struct in6_addr)) == 0) {
+			iface = net_if_ipv6_select_src_iface(&addr6->sin6_addr);
+			net_context_set_iface(context, iface);
+		}
+
```
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/ip/net_context.c:1472]: yep, will fix

**Comment by hubertmis** [comment]
Now I'm thinking about potential functional change with this patch. Let's assume the device has two network interfaces A with IP address `fd00::1/64` and B with IP address `fd01::5/64`. The application creates an UDP socket and binds it to address `fd00::1`, the context interface is set to A. The application receives a frame with destination address `fd00::1` and source address `fd01::2`. The application sends a response with destination address `fd01::2` and source address `fd00::1`.

I think that without this patch the response would be sent through interface A, but with this patch interface B would be selected.

I think that proposed this change is correct (the output interface for response should be selected by the destination address and routing rules, not matching the interface from which request was received). However I would like to verify if my understanding is correct and if this change intentional.

**Comment by jukkar** [comment]
> The application creates an UDP socket and binds it to address `fd00::1`

Note that the patch only handles the case where the application has not bound to specific src address. Your use case should work just fine. After some more testing, the issue seems to be more related to IPv4 which does not work properly without this patch.

**Comment by hubertmis** [comment]
> Note that the patch only handles the case where the application has not bound to specific src address

Does it? This patch checks remote address of a net context, while `bind()` function sets local address. My understanding is that it applies to sockets that are bound, but not connected.

**Comment by jukkar** [comment]
>  This patch checks remote address of a net context, while `bind()` function sets local address. My understanding is that it applies to contexts that are not connected, including bound contexts.

Sure, but we do a default bind if user has not done it already in order to select the source interface. If user has not bound, then `context->remote` is not set which then causes problems when network interface is selected later. Setting the interface here helps later in the code.

**Comment by hubertmis** [comment]
@jukkar, forgive me for digging deeper in this patch, but I would like to seize the opportunity to learn more about network stack while I'm reviewing.

> If user has not bound, then `context->remote` is not set

But if user has bound, `context->remote` is not set as well. Or am I missing something?

> Setting the interface here helps later in the code.

I agree and I have not marked -1 because of this. However, I'm interested in side effects of this change and I would like to verify if my understanding of network components, I've been gaining recently, is correct.

I don't plan to mark this PR as +1, because I don't feel competent enough. I think other reviewers should take a look as well.

**Comment by jukkar** [comment]
> But if user has bound, `context->remote` is not set as well. Or am I missing something?

No you are not missing anything. I seem to have used wrong wording, I will revisit this PR tomorrow, it has been a long day and need to some rest.

**Comment by jukkar** [comment]
I clarified the comments and commit message. Basically when I was talking about binding, I should have said connecting. So as Hubert said, the `context->remote` is related to target address which could have been set by `connect()` call.
So if we have not called `connect()`, then when sending a packet (UDP in this case), the target network interface might be missing when we are figuring out the local address to use (default bind if user has not bound it, which is typically the case with UDP and sendto()).

**Comment by hubertmis** [comment]
Thanks for the explanation. Now I think I understand what's going on.

I'm thinking if this change should be broader. I mean, for each packet being sent the output interface should be selected when the packet is added to TX queue, not when the socket is connected. In current implementation, when the socket is connected one interface is selected for this socket based on some rules. However, the reasons to use particular interface to reach given destination may change (network addresses, routing topology) while socket is connected. If such reasons change, packet sent through given socket should start using other interface than previously, what seems missing in current code.

It's obviously out of scope of this PR. But maybe we could start thinking of network stack refactoring, to easily support devices which from network point of view are routers with dynamic routing tables.

**Comment by jukkar** [comment]
I still miss one +1, anyone?

**Review by mniestroj — APPROVED** [review]

**Pull request merged**